### PR TITLE
Base functionality for reassembling a traversed sequence

### PR DESCRIPTION
**Pull request by saber3005**
I wrote the base functionality and have done a little bit of manual testing. If you like how this looks, I'll add some unit tests.

### CODE REVIEW

File: src/IterationTraits.php [hunk 1]
```diff
@@ -299,4 +299,33 @@ public static function transpose(Iterator $iterator) {
...
+     * Reassembles a traversed sequence into its original shape
+     *
+     * @param Iterator $iterator
+     * @return Sequence
+     */
+    public static function reassemble(Iterator $iterator) {
+        return self::wrapFunctionIntoSequenceOnDemand(function() use ($iterator) {
+           return Sequence::make($iterator)
+               ->reduceToSequence(array(), function ($collection, $value, $path) {
+                   $keys = explode('.', $path);
```
Comment (Jason-Rev) [inline src/IterationTraits.php]: We have a getter/setter library that will inject the values into an array.
`use Revinate\GetterSetter as gs;`

`
return gs\set($collection, $path, $value);
`

**Comment by pbu88** [comment]
Can you provide a basic use case for what you need this functionality, and maybe some example of input output?

File: src/IterationTraits.php (hunk 1 above)
Comment (saber3005) [inline src/IterationTraits.php]: I've made that change in commit 19b6d832fb04eca60d594d7572d8c47c2a0f7f0c

**Comment by saber3005** [comment]
@pbu88 The specific use case that I have is parsing all of the values of a YAML file to implement a DSL that we're writing.

Here's a basic code example:
```php
$array = array('foo' => 2, 'bar' => array('baz' => 6, 'qux' => 5));
$doubled = Sequence::make($array)->traverse()
->map(function ($val) {
    return $val*2;
})->reassemble()->toArray();

$doubled === array('foo' => 4, 'bar' => array('baz' => 12, 'qux' => 10));
```

Doing this with traverse allows us to map every value in an array tree, regardless of how deep it is. The only problem as of right now is that there's no easy way to get it all packaged back up in the original data structure.

**Comment by Jason-Rev** [comment]
Thank you Eric.  If you make some unit tests, then I think we are good to go.

**Comment by pbu88** [comment]
Thank you for the use case scenario. I think is a valid function to have too but on the other side, I think we might want a __recursive__ `walk()` instead, because from what I see this is the perfect use case for that. It's more efficient and more natural.

Also, I see little use for `reassemble()` by itself (without `traverse()` being called first). Maybe we can start questioning the possibility to add a `contrib` namespace where we can put these useful functions that don't really belong to the core. This way we can keep the core small, readable and concise.

Can you guys @saber3005  and @Jason-Rev comment on this?

**Comment by Jason-Rev** [comment]
Hey Paulo,

In principle, I think you are right, this type of problem would be best solved with the ability to traverse into an array of documents and apply a map to a selected set of fields.  We have discussed ways of doing this, but that implementation is beyond the scope of this ask.

At the same time, we do provide traverse, but do not provide the dual functionality.  `reassemble()` provides that ability.  My only add would be to enable other separators.

I agree with you about trying to keep the number of functions within Sequence to a minimum.

**Comment by saber3005** [comment]
@Jason-Rev, I've updated my changes to add support for alternative separators to both `traverse` and `reassemble` and to add unit tests for all new functionality.

@pbu88 I also agree with minimizing the number of functions in Sequence. I would say that dedicated recursive walk/map/filter functions would be very useful, though for now having full support for traverse/reassemble allows anyone to approximate that functionality along with any other use they might have. In fact, you could probably implement placeholder recursive walk/map/filter functions by using eg. `traverse()->map()->reassemble()` while you decide on a more efficient implementation.

**Pull request merged**